### DISCUSSION TITLE
Dependent tasks with atoms

### DISCUSSION THREAD

**Discussion by valentincostam** [pmndrs/jotai#1401] [Q&A] [answered]
On this project, tasks can depend on other tasks' state. _If a task's requirements are not met, it should be disabled._

The tasks are always the same. You can't add new ones nor delete them. **Only their state can change.**

Each task's state is kept in an `atom`.

Given the following array of task objects:

```js
const tasks = [
  { id: 1, name: "Get milk", state: atom("pending") },
  { id: 2, name: "Get coffee", state: atom("pending") },
  {
    id: 3,
    name: "Make cappuccino",
    state: atom("pending"),
    requirements: [
      { id: "r1", requiredTaskId: 1, requiredState: "ongoing" },
      { id: "r2", requiredTaskId: 2, requiredState: "ongoing" }
    ]
  },
  ...
];
```

I hoped to do this:

```js
const Task = ({ task }) => {
  const [state, setState] = useAtom(task.state);

  // What I would try to do:
  let isEnabled = true;

  if (task.requirements?.length) {
    isEnabled = task.requirements.every(req => {
      const requiredTask = tasks.find(task => task.id === req.requiredTaskId)
      const [requiredTaskState] = useAtom(requiredTask.state)
      return requiredTaskState === req.requiredState
    })
  }

  return (
    <li>
      <select
        value={state}
        onChange={(e) => setState(e.target.value)}
        disabled={!isEnabled}
      >
        ...
      </select>
      ...
    </li>
  );
};
```

But I get a `React Hook "useAtom" cannot be called inside a callback` warning (as I expected) and a `Rendered more hooks than during the previous render` error when I change some task state.

**How can I get requirements' state inside the `Task` component?** What is the right approach in this case?

Please, check this example on CodeSandbox: [https://codesandbox.io/s/jotai-dependant-tasks-nvii23](https://codesandbox.io/s/jotai-dependant-tasks-nvii23)

First time using Jotai and I'm not an experienced React developer.

Thanks in advance.

**Comment by dai-shi** [comment, marked as answer]
This is a little tricky, but something like this would work:
```js
const Task = ({ task }) => {
  const [state, setState] = useAtom(task.state);

  const [isEnabled] = useAtom(useMemo(() => atom((get) => task.requirements.every((req) => {
    const requiredTask = tasks.find(task => task.id === req.requiredTaskId)
    const requiredTaskState = get(requiredTask.state)
    return requiredTaskState === req.requiredState
  })), [...])); // follow eslint react-hooks rules

```

**Reply by valentincostam** [reply]
Thanks a lot for the quick response, Daishi. That's exactly what I was looking for. ❤️ 

Now the example is working as you can see in [this sandbox](https://codesandbox.io/s/jotai-dependant-tasks-forked-nl9kbr). 

**When you say "tricky", do you mean "_the code is a little hard to read_"? Or "_maybe Jotai is not the best solution for this_"?**

I tried to make a little refactor without luck.

Instead of getting `isEnabled` in the component definition like this (as you recommended):

```js
const Task = ({ task }) => {
  const [state, setState] = useAtom(task.state);

  const [isEnabled] = useAtom(
    useMemo(
      () =>
        atom((get) =>
          task.requirements?.every((req) => {
            const requiredTask = tasks.find((task) => task.id === req.requiredTaskId);
            const requiredTaskState = get(requiredTask.state);
            return [req.requiredState, "completed"].includes(requiredTaskState);
          })
        ),
      [task.requirements]
    )
  );

  ...
}
```

I wanted to create an atom here (attempt no. 1):

```js
const tasks = [
  ...
  {
    id: 3,
    name: "Make cappuccino",
    stage: 2,
    state: atom("pending"),
    requirements: [
      { id: "r1", requiredTaskId: 1, requiredState: "ongoing" },
      { id: "r2", requiredTaskId: 2, requiredState: "ongoing" }
    ],
    // EDIT: Later I realized this getter will create a new atom each time is called, right?
    get isEnabled() {
      return atom((get) => {
        this.requirements.every((req) => {
          const requiredTask = tasks.find((task) => task.id === req.requiredTaskId);
          const requiredTaskState = get(requiredTask.state);
          return [req.requiredState, "completed"].includes(requiredTaskState);
        });
      });
    },
  },
  ...
]
```

Or even like this (attempt no. 2):

```js
tasks.forEach((task) => {
  task.isEnabled = atom((get) => {
    task.requirements.every((req) => {
      const requiredTask = tasks.find((task) => task.id === req.requiredTaskId);
      const requiredTaskState = get(requiredTask.state);
      return [req.requiredState, "completed"].includes(requiredTaskState);
    });
  });
});
```

So I could get `isEnabled` this way:

```js
const [isEnabled] = useAtom(useMemo(() => task.isEnabled, [task.isEnabled]));
```

But that's definitely wrong since it produces an infinite loop in one (or both?) of the attempts. 

**Do you think there is a better way to define that atom (`isEnabled`) for each task?**

Since it works with the lines you suggested, I'm already pretty happy with it.

That said, on my closer-to-real-use-case example ([see here](https://codesandbox.io/s/jotai-dependant-tasks-forked-op2qvx)), I have other atoms similar to that you proposed (see `isCompleted` and `canComplete` in `Stage` component). So I would like to keep those as simple and readable as I can too.

Maybe it's just as simple as creating some functions for `every` and `filter`'s callbacks, moving some logic out of the components. But I'm trying to see if there is a better way. A clean, right, Jotai way?

Thanks again for taking the time to reply, Daishi.

Jotai is awesome, and it's such a privilege to get help from you. 🙏

**Reply by dai-shi** [reply]
> When you say "tricky", do you mean "the code is a little hard to read"? Or "maybe Jotai is not the best solution for this"?

The former. It's a valid pattern and recommended if we need to make atoms depend on props. But, for jotai beginners, it would be hard to read and notice such practices.

> Do you think there is a better way to define that atom (isEnabled) for each task?

Yes, I'd recommend the atoms-in-atom pattern.

```js
const tasksAtom = atom([
  ...
  {
    id: 3,
    name: "Make cappuccino",
    stage: 2,
    state: atom("pending"),
    requirements: [
      { id: "r1", requiredTaskId: 1, requiredState: "ongoing" },
      { id: "r2", requiredTaskId: 2, requiredState: "ongoing" }
    ],
    isEnabled: atom((get) => {
      const tasks = get(tasksAtom)
      const me = tasks.find(({ id }) => id === 3) // hard coded in this example
      return me.requirements.every((req) => {
        const requiredTask = tasks.find((task) => task.id === req.requiredTaskId);
        const requiredTaskState = get(requiredTask.state);
        return [req.requiredState, "completed"].includes(requiredTaskState);
      });
    }),
  },
  ...
])
```

You can make a helper function to simplify.
